File: crates/riptide-api/src/handlers/extract.rs
//! Content extraction handler for single-URL extraction
//!
//! Provides a unified endpoint for extracting content from URLs using
//! the multi-strategy extraction pipeline and riptide-facade.

use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};
use std::time::Instant;

use crate::context::ApplicationContext;

// Import HTTP DTOs from riptide-types (Phase 2C.1 - breaking circular dependency)
use riptide_types::{ExtractRequest, ExtractionMethod};

/// Parse strategy string to ExtractionMethod enum
pub(crate) fn parse_extraction_strategy(strategy_str: &str) -> Option<ExtractionMethod> {
    match strategy_str.to_lowercase().as_str() {
        "native" | "css" => Some(ExtractionMethod::HtmlCss),
        "wasm" => Some(ExtractionMethod::Wasm),
        "regex" => Some(ExtractionMethod::HtmlRegex),
        "fallback" => Some(ExtractionMethod::Fallback),
        "auto" | "multi" => None, // None means use default UnifiedExtractor
        "markdown" => None,       // Handled separately via as_markdown flag
        _ => None,
    }
}

/// Extract content from a URL using multi-strategy extraction
///
/// This endpoint provides a unified interface for content extraction,
/// delegating to the ExtractionFacade for all business logic.
#[axum::debug_handler]
#[tracing::instrument(skip(state), fields(url = %payload.url, mode = %payload.mode))]
pub async fn extract(
    State(state): State<ApplicationContext>,
    Json(payload): Json<ExtractRequest>,
) -> impl IntoResponse {
    let start = Instant::now();

    // Validate URL format (HTTP concern)
    if let Err(e) = url::Url::parse(&payload.url) {
        tracing::warn!(url = %payload.url, error = %e, "Invalid URL format");
        return crate::errors::ApiError::invalid_url(&payload.url, e.to_string()).into_response();
    }

    // Parse strategy from request
    let extraction_strategy = parse_extraction_strategy(&payload.options.strategy);

    // Build extraction options from request
    let html_options = riptide_facade::facades::HtmlExtractionOptions {
        as_markdown: payload.options.strategy == "markdown",
        clean: true,
        include_metadata: true,
        extract_links: false,
        extract_images: false,
        custom_selectors: None,
        extraction_strategy,
    };

    // Delegate to facade (handles fetch + extraction)
    match state
        .extraction_facade
        .extract_from_url(&payload.url, html_options)
        .await
    {
        Ok(extracted) => {
            // Include raw HTML only if explicitly requested
            let raw_html = if payload.options.include_html {
                extracted.raw_html.clone()
            } else {
                None
            };

            let response = riptide_types::ExtractResponse {
                url: extracted.url.clone(),
                title: extracted.title.clone(),
                content: extracted.text.clone(),
                metadata: riptide_types::ContentMetadata {
                    author: extracted.metadata.get("author").cloned(),
                    publish_date: extracted.metadata.get("publish_date").cloned(),
                    word_count: extracted.text.split_whitespace().count(),
                    language: extracted.metadata.get("language").cloned(),
                },
                strategy_used: extracted.strategy_used,
                quality_score: extracted.confidence,
                extraction_time_ms: start.elapsed().as_millis() as u64,
                parser_metadata: None,
                raw_html,
            };
            (StatusCode::OK, Json(response)).into_response()
        }
        Err(e) => crate::errors::ApiError::from(e).into_response(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_request_deserialization() {
        let json = r#"{"url": "https://example.com"}"#;
        let req: ExtractRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.url, "https://example.com");
        assert_eq!(req.mode, "standard");
        assert_eq!(req.options.strategy, "multi");
    }

    #[test]
    fn test_extract_request_with_options() {
        let json = r#"{
            "url": "https://example.com",
            "mode": "article",
            "options": {
                "strategy": "css",
                "quality_threshold": 0.8,
                "timeout_ms": 10000
            }
        }"#;
        let req: ExtractRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.mode, "article");
        assert_eq!(req.options.strategy, "css");
        assert_eq!(req.options.quality_threshold, 0.8);
        assert_eq!(req.options.timeout_ms, 10000);
    }
}

File: sdk/python/riptide_sdk/endpoints/extract.py
"""
Extract API endpoint implementation

Provides single-URL content extraction with multi-strategy support including
CSS, WASM, and hybrid extraction pipelines.
"""

from typing import Optional
import httpx

from ..models import ExtractOptions, ExtractionResult
from ..exceptions import APIError, ValidationError


class ExtractAPI:
    """API for content extraction operations"""

    def __init__(self, client: httpx.AsyncClient, base_url: str):
        """
        Initialize ExtractAPI

        Args:
            client: Async HTTP client
            base_url: Base URL for the API
        """
        self.client = client
        self.base_url = base_url

    async def extract(
        self,
        url: str,
        mode: str = "standard",
        options: Optional[ExtractOptions] = None,
    ) -> ExtractionResult:
        """
        Extract content from a single URL using multi-strategy extraction

        This endpoint provides a unified interface for content extraction,
        using the multi-strategy extraction pipeline (CSS, WASM, fallback).

        Args:
            url: URL to extract content from
            mode: Extraction mode - "standard", "article", "product", or "markdown"
            options: Optional extraction configuration

        Returns:
            ExtractionResult with extracted content and metadata

        Raises:
            ValidationError: If URL is invalid
            APIError: If the API returns an error

        Example:
            >>> # Basic extraction
            >>> result = await client.extract.extract("https://example.com")
            >>> print(f"Title: {result.title}")
            >>> print(f"Content: {result.content[:100]}...")
            >>> print(f"Quality: {result.quality_score:.2f}")
            >>>
            >>> # With custom options
            >>> from riptide_sdk.models import ExtractOptions
            >>> options = ExtractOptions(
            ...     strategy="wasm",
            ...     quality_threshold=0.8,
            ...     timeout_ms=15000
            ... )
            >>> result = await client.extract.extract(
            ...     "https://example.com/article",
            ...     mode="article",
            ...     options=options
            ... )
            >>> print(f"Strategy used: {result.strategy_used}")
            >>> print(f"Word count: {result.metadata.word_count}")
        """
        # Validate URL
        if not url:
            raise ValidationError("URL cannot be empty")

        if not url.startswith(("http://", "https://")):
            raise ValidationError(f"Invalid URL protocol: {url}")

        # Build request body
        body = {
            "url": url,
            "mode": mode,
        }

        if options:
            body["options"] = options.to_dict()

        # Make request
        response = await self.client.post(
            f"{self.base_url}/api/v1/extract",
            json=body,
        )

        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            raise APIError(
                message=error_data.get("error", {}).get("message", "Extraction failed"),
                status_code=response.status_code,
                response_data=error_data,
            )

        return ExtractionResult.from_dict(response.json())

    async def extract_article(
        self,
        url: str,
        options: Optional[ExtractOptions] = None,
    ) -> ExtractionResult:
        """
        Extract content in article mode (optimized for news/blog posts)

        Args:
            url: URL to extract content from
            options: Optional extraction configuration

        Returns:
            ExtractionResult with article content

        Example:
            >>> result = await client.extract.extract_article(
            ...     "https://example.com/blog/post"
            ... )
            >>> print(f"Author: {result.metadata.author}")
            >>> print(f"Published: {result.metadata.publish_date}")
        """
        return await self.extract(url, mode="article", options=options)

    async def extract_markdown(
        self,
        url: str,
        options: Optional[ExtractOptions] = None,
    ) -> ExtractionResult:
        """
        Extract content as Markdown

        Args:
            url: URL to extract content from
            options: Optional extraction configuration

        Returns:
            ExtractionResult with content in Markdown format

        Example:
            >>> result = await client.extract.extract_markdown(
            ...     "https://example.com"
            ... )
            >>> print(result.content)  # Markdown formatted
        """
        return await self.extract(url, mode="markdown", options=options)

    async def extract_product(
        self,
        url: str,
        options: Optional[ExtractOptions] = None,
    ) -> ExtractionResult:
        """
        Extract content in product mode (optimized for e-commerce)

        Args:
            url: URL to extract content from
            options: Optional extraction configuration

        Returns:
            ExtractionResult with product information

        Example:
            >>> result = await client.extract.extract_product(
            ...     "https://shop.example.com/product/123"
            ... )
            >>> print(f"Product: {result.title}")
        """
        return await self.extract(url, mode="product", options=options)
